src/wampy.js

```diff
@@ -1767,57 +1767,45 @@ class Wampy {
 
     /**
      * Unsubscribe from topic
-     * @param {string|number} subscriptionIdKey Subscription ID or Key, received during .subscribe()
+     * @param {string|number} subscriptionIdOrKey Subscription ID or Key, received during .subscribe()
      * @param {function} [onEvent] - received event callback to remove (optional). If not provided -
      *                               all callbacks will be removed and unsubscribed on the server
      * @returns {Promise}
      */
-    async unsubscribe (subscriptionIdKey, onEvent) {
-        let reqId;
-        const callbacks = getNewPromise();
-
+    async unsubscribe (subscriptionIdOrKey, onEvent) {
         if (!this._preReqChecks(null, 'broker')) {
             throw this._cache.opStatus.error;
         }
 
-        const sub = this._subscriptionsById.get(subscriptionIdKey) ||
-            this._subscriptionsByKey.get(subscriptionIdKey);
-        if (sub) {
+        const subscription = this._subscriptionsById.get(subscriptionIdOrKey) ||
+            this._subscriptionsByKey.get(subscriptionIdOrKey);
 
-            reqId = this._getReqId();
+        if (!subscription) {
+            const nonExistUnsubscribeError = new Errors.NonExistUnsubscribeError();
+            this._fillOpStatusByError(nonExistUnsubscribeError);
+            throw nonExistUnsubscribeError;
+        }
 
-            if (typeof onEvent === 'function') {
-                const i = sub.callbacks.indexOf(onEvent);
-                if (i >= 0) {
-                    sub.callbacks.splice(i, 1);
-                }
-            } else {
-                sub.callbacks = [];
-            }
+        subscription.callbacks = typeof onEvent === 'function'
+            ? subscription.callbacks = subscription.callbacks.filter((callback) => callback !== onEvent)
```

Comment:
```suggestion
            ? subscription.callbacks.filter((callback) => callback !== onEvent)
```

Reply:
Seems that's not needed.

Reply:
The previous solution was using Array.splice() to mutate the original array, however Array.filter() method doesn't mutate the original array, it creates a [shallow copy](https://developer.mozilla.org/en-US/docs/Glossary/Shallow_copy) of it. So I had to add this in order to keep mutating the `subscription.callbacks` array, so it has the same behavior as before.